plugin/firewall/firewall.go

```diff
@@ -0,0 +1,126 @@
+// Package firewall enables filtering on query and response using direct expression as policy.
+// it allows interact with other Policy Engines if those are plugin implementing the Engineer interface
+package firewall
+
+import (
+	"context"
+	"errors"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/firewall/policy"
+	"github.com/coredns/coredns/plugin/firewall/rule"
+	"github.com/coredns/coredns/plugin/pkg/dnstest"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+var (
+	errInvalidAction = errors.New("invalid action")
+)
+
+// ExpressionEngineName is the name associated with built-in rules of Expression type.
+const ExpressionEngineName = "--default--"
+
+// firewall represents a plugin instance that can validate DNS
+// requests and replies using rulelists on the query and/or on the reply
+type firewall struct {
+	engines map[string]policy.Engine
+	query   *rule.List
+	reply   *rule.List
+
+	next plugin.Handler
+}
+
+//New build a new firewall plugin
+func New() (*firewall, error) {
+	pol := &firewall{engines: map[string]policy.Engine{"--default--": policy.NewExprEngine()}}
+	var err error
+	if pol.query, err = rule.NewList(policy.TypeBlock, false); err != nil {
+		return nil, err
+	}
+	if pol.reply, err = rule.NewList(policy.TypeAllow, true); err != nil {
+		return nil, err
+	}
+	return pol, nil
+}
+
+// ServeDNS implements the Handler interface.
+func (p *firewall) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
```

Comment:
this method doesn't check if it is authoritative for the domains it is seeing.

Reply:
hum ... there is no ZONE in the directive .. should we add it ?

Reply:
if we add, then should we consider that we could have different filter depending the zone ?
something like

```
firewall query google.com {
...
}
firewall query azure.com {
...
}
firewall query . {
...
}

```